From 13eea0d3f8800426f58d4ac517b77eb597882692 Mon Sep 17 00:00:00 2001
From: Volodymyr Samokhatko <volodymyr.samokhatko@gmail.com>
Date: Wed, 5 Apr 2017 22:52:27 +0200
Subject: [PATCH 241/254] Fix populating selection clipboard with keyboard

Task-number: QTBUG-59879
Pick-to: 6.0 5.15
Change-Id: I6948919fc90995c60a34b5bd6b4a225c1a59fd9b
Reviewed-by: Liang Qi <liang.qi@qt.io>
(cherry picked from commit 976f28e7bb7cddb11459600f293868abcf7da948)
---
 src/widgets/widgets/qplaintextedit.cpp     |  2 +-
 src/widgets/widgets/qtextedit.cpp          |  2 +-
 src/widgets/widgets/qwidgetlinecontrol.cpp |  9 +++++++--
 src/widgets/widgets/qwidgettextcontrol.cpp | 14 +++++++++++++-
 src/widgets/widgets/qwidgettextcontrol_p.h |  2 +-
 5 files changed, 23 insertions(+), 6 deletions(-)

diff --git a/src/widgets/widgets/qplaintextedit.cpp b/src/widgets/widgets/qplaintextedit.cpp
index 87e8af1382..6694b79dda 100644
--- a/src/widgets/widgets/qplaintextedit.cpp
+++ b/src/widgets/widgets/qplaintextedit.cpp
@@ -972,7 +972,7 @@ void QPlainTextEditPrivate::pageUpDown(QTextCursor::MoveOperation op, QTextCurso
     }
 
     if (moveCursor) {
-        control->setTextCursor(cursor);
+        control->setTextCursor(cursor, moveMode == QTextCursor::KeepAnchor);
         pageUpDownLastCursorYIsValid = true;
     }
 }
diff --git a/src/widgets/widgets/qtextedit.cpp b/src/widgets/widgets/qtextedit.cpp
index 8ddda78f7d..656062fcd6 100644
--- a/src/widgets/widgets/qtextedit.cpp
+++ b/src/widgets/widgets/qtextedit.cpp
@@ -270,7 +270,7 @@ void QTextEditPrivate::pageUpDown(QTextCursor::MoveOperation op, QTextCursor::Mo
             vbar->triggerAction(QAbstractSlider::SliderPageStepAdd);
         }
     }
-    control->setTextCursor(cursor);
+    control->setTextCursor(cursor, moveMode == QTextCursor::KeepAnchor);
 }
 
 #if QT_CONFIG(scrollbar)
diff --git a/src/widgets/widgets/qwidgetlinecontrol.cpp b/src/widgets/widgets/qwidgetlinecontrol.cpp
index 9dd61c2c6a..087657ab85 100644
--- a/src/widgets/widgets/qwidgetlinecontrol.cpp
+++ b/src/widgets/widgets/qwidgetlinecontrol.cpp
@@ -1948,10 +1948,15 @@ void QWidgetLineControl::processKeyEvent(QKeyEvent* event)
         return;
     }
 
-    if (unknown)
+    if (unknown) {
         event->ignore();
-    else
+    } else {
+#ifndef QT_NO_CLIPBOARD
+        if (QApplication::clipboard()->supportsSelection())
+            copy(QClipboard::Selection);
+#endif
         event->accept();
+    }
 }
 
 bool QWidgetLineControl::isUndoAvailable() const
diff --git a/src/widgets/widgets/qwidgettextcontrol.cpp b/src/widgets/widgets/qwidgettextcontrol.cpp
index e2a07c0043..961f954b95 100644
--- a/src/widgets/widgets/qwidgettextcontrol.cpp
+++ b/src/widgets/widgets/qwidgettextcontrol.cpp
@@ -922,7 +922,7 @@ QTextDocument *QWidgetTextControl::document() const
     return d->doc;
 }
 
-void QWidgetTextControl::setTextCursor(const QTextCursor &cursor)
+void QWidgetTextControl::setTextCursor(const QTextCursor &cursor, bool selectionClipboard)
 {
     Q_D(QWidgetTextControl);
     d->cursorIsFocusIndicator = false;
@@ -936,6 +936,11 @@ void QWidgetTextControl::setTextCursor(const QTextCursor &cursor)
     d->repaintOldAndNewSelection(oldSelection);
     if (posChanged)
         emit cursorPositionChanged();
+
+#ifndef QT_NO_CLIPBOARD
+    if (selectionClipboard)
+        d->setClipboardSelection();
+#endif
 }
 
 QTextCursor QWidgetTextControl::textCursor() const
@@ -1225,6 +1230,9 @@ void QWidgetTextControlPrivate::keyPressEvent(QKeyEvent *e)
     if (e == QKeySequence::SelectAll) {
             e->accept();
             q->selectAll();
+#ifndef QT_NO_CLIPBOARD
+            setClipboardSelection();
+#endif
             return;
     }
 #ifndef QT_NO_CLIPBOARD
@@ -1376,6 +1384,10 @@ process:
 
  accept:
 
+#ifndef QT_NO_CLIPBOARD
+    setClipboardSelection();
+#endif
+
     e->accept();
     cursorOn = true;
 
diff --git a/src/widgets/widgets/qwidgettextcontrol_p.h b/src/widgets/widgets/qwidgettextcontrol_p.h
index c445ecaf80..71dc988b56 100644
--- a/src/widgets/widgets/qwidgettextcontrol_p.h
+++ b/src/widgets/widgets/qwidgettextcontrol_p.h
@@ -105,7 +105,7 @@ public:
     void setDocument(QTextDocument *document);
     QTextDocument *document() const;
 
-    void setTextCursor(const QTextCursor &cursor);
+    void setTextCursor(const QTextCursor &cursor, bool selectionClipboard = false);
     QTextCursor textCursor() const;
 
     void setTextInteractionFlags(Qt::TextInteractionFlags flags);
-- 
2.33.1

